File: todo-api-rs/src/domain/entities/todo.rs
use std::fmt;

use thiserror::Error;

use crate::domain::types::{Date, DateTime, Id};

#[derive(Clone, Debug)]
pub struct TodoEntity {
    id: Id,
    pub title: Title,
    pub description: Option<Description>,
    pub status: Status,
    pub todo_at: Option<Date>,
    created_at: Option<DateTime>,
    updated_at: Option<DateTime>,
}

impl TodoEntity {
    pub fn new(props: NewProps) -> Self {
        Self {
            id: Id::new(),
            title: props.title,
            description: props.description,
            status: props.status,
            todo_at: props.todo_at,
            created_at: None,
            updated_at: None,
        }
    }

    pub fn init(props: InitProps) -> Self {
        Self {
            id: props.id,
            title: props.title,
            description: props.description,
            status: props.status,
            todo_at: props.todo_at,
            created_at: props.created_at,
            updated_at: props.updated_at,
        }
    }

    pub fn id(&self) -> Id {
        self.id
    }

    pub fn created_at(&self) -> Option<DateTime> {
        self.created_at
    }

    pub fn updated_at(&self) -> Option<DateTime> {
        self.updated_at
    }
}

#[derive(Clone, Debug)]
pub struct NewProps {
    pub title: Title,
    pub description: Option<Description>,
    pub status: Status,
    pub todo_at: Option<Date>,
}

#[derive(Clone, Debug)]
pub struct InitProps {
    pub id: Id,
    pub title: Title,
    pub description: Option<Description>,
    pub status: Status,
    pub todo_at: Option<Date>,
    pub created_at: Option<DateTime>,
    pub updated_at: Option<DateTime>,
}

impl PartialEq for TodoEntity {
    fn eq(&self, other: &Self) -> bool {
        self.id == other.id
    }
}

impl Eq for TodoEntity {}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Title(String);

impl Title {
    pub const MAX_LENGTH: usize = 64;

    pub fn new(title: impl Into<String>) -> Result<Self, TitleError> {
        let title: String = title.into();
        if title.is_empty() {
            return Err(TitleError::Empty);
        } else if title.len() > Self::MAX_LENGTH {
            return Err(TitleError::Length);
        }

        Ok(Self(title))
    }

    pub fn into_inner(self) -> String {
        self.0
    }

    pub fn as_str(&self) -> &str {
        self.0.as_str()
    }
}

impl fmt::Display for Title {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.0.fmt(f)
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct Description(String);

impl Description {
    pub const MAX_LENGTH: usize = 256;

    pub fn new(description: impl Into<String>) -> Result<Self, DescriptionError> {
        let description: String = description.into();
        if description.len() > Self::MAX_LENGTH {
            return Err(DescriptionError::Length);
        }

        Ok(Self(description))
    }

    pub fn into_inner(self) -> String {
        self.0
    }

    pub fn as_str(&self) -> &str {
        &self.0
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum Status {
    Todo,
    InProgress,
    Done,
}

impl Status {
    const TODO_STR: &'static str = "todo";
    const IN_PROGRESS_STR: &'static str = "in_progress";
    const DONE_STR: &'static str = "done";

    pub fn parse_str(value: &str) -> Result<Self, StatusError> {
        match value {
            Self::TODO_STR => Ok(Self::Todo),
            Self::IN_PROGRESS_STR => Ok(Self::InProgress),
            Self::DONE_STR => Ok(Self::Done),
            _ => Err(StatusError),
        }
    }
}

impl fmt::Display for Status {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Todo => f.write_str(Self::TODO_STR),
            Self::InProgress => f.write_str(Self::IN_PROGRESS_STR),
            Self::Done => f.write_str(Self::DONE_STR),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Error)]
pub enum TitleError {
    #[error("Todo title cannot be empty")]
    Empty,
    #[error("Todo title cannot have more than 64 characters")]
    Length,
}

#[derive(Clone, Debug, PartialEq, Eq, Error)]
pub enum DescriptionError {
    #[error("Todo description cannot have more than 64 characters")]
    Length,
}

#[derive(Clone, Debug, PartialEq, Eq, Error)]
#[error(
    "Todo status must be one the following values: {}, {}, {}",
    Status::TODO_STR,
    Status::IN_PROGRESS_STR,
    Status::DONE_STR
)]
pub struct StatusError;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_title_fails() {
        let empty_title = Title::new("");
        assert_eq!(empty_title, Err(TitleError::Empty));
    }

    #[test]
    fn title_too_big_fails() {
        let src =
            "This is a really big title so it should not satisfy title max length constraints";
        let big_title = Title::new(src);
        assert!(src.len() > Title::MAX_LENGTH);
        assert_eq!(big_title, Err(TitleError::Length));
    }

    #[test]
    fn new_title_works() {
        let src = "Title";
        let title = Title::new(src);
        assert!(title.is_ok());
        assert_eq!(Ok(src), title.as_ref().map(Title::as_str))
    }

    #[test]
    fn title_formats_to_string() {
        let src = String::from("Title");
        let title = Title::new(&src);
        assert_eq!(Ok(src.as_str()), title.as_ref().map(Title::as_str));
        assert_eq!(Ok(src), title.as_ref().map(Title::to_string));
    }
}

File: todo-api-js/src/domain/entities/todo.ts
import { Id } from "@domain/value-objects/id";
import { Ok, Err, type Result } from "@domain/utils/result";
import { ValueObject } from "@domain/value-objects";
import { Entity } from ".";

export type TodoEntityProps = {
    id: Id;
    title: Title;
    description?: Description | undefined;
    status: Status;
    todoAt?: Date | undefined;
    createdAt: Date;
    updatedAt: Date;
};

export type CreateTodoEntityProps = {
    title: Title;
    description?: Description | undefined;
    status: Status;
    todoAt?: Date | undefined;
};

export class TodoEntity extends Entity<TodoEntityProps> {
    static create(props: CreateTodoEntityProps): TodoEntity {
        const now = new Date();
        return new TodoEntity({ ...props, id: Id.create(), createdAt: now, updatedAt: now });
    }

    static from(props: Readonly<TodoEntityProps>): TodoEntity {
        return new TodoEntity(props);
    }

    get id(): Id {
        return this.props.id;
    }

    get title(): Title {
        return this.props.title;
    }
}

export class Title extends ValueObject<Title> {
    static readonly MIN_LENGTH = 3;
    static readonly MAX_LENGTH = 64;

    private constructor(public readonly value: string) {
        super();
    }

    static parse(title: string): Result<Title, string> {
        if (title.length > Title.MAX_LENGTH) {
            const message = `Todo title cannot have more than ${Title.MAX_LENGTH} characters`;
            return new Err(message);
        }

        if (title.length < Title.MIN_LENGTH) {
            const message = `Todo title must have at least ${Title.MIN_LENGTH} characters`;
            return new Err(message);
        }

        return new Ok(new Title(title));
    }

    eq(rhs: Title): boolean {
        return this.value === rhs.value;
    }
}

export class Description extends ValueObject<Description> {
    static readonly MAX_LENGTH = 128;

    private constructor(public readonly value: string) {
        super();
        this.value = value;
    }

    static parse(description: string): Result<Description, string> {
        if (description.length > this.MAX_LENGTH) {
            const message = `Todo description cannot have more than ${this.MAX_LENGTH} characters`;
            return new Err(message);
        }
        return new Ok(new Description(description));
    }

    eq(rhs: Description): boolean {
        return this.value === rhs.value;
    }
}

export type StatusKind = "todo" | "in_progress" | "done";

export class Status extends ValueObject<Status> {
    static readonly OPTIONS: StatusKind[] = ["todo", "in_progress", "done"];

    constructor(public readonly value: StatusKind) {
        super();
    }

    static parse(value: string): Result<Status, string> {
        if (!Status.OPTIONS.includes(value as StatusKind)) {
            const options = Status.OPTIONS.join(", ");
            const message = `Todo status must be one of the following options: ${options}`;
            return new Err(message);
        }
        return new Ok(new Status(value as StatusKind));
    }

    eq(rhs: Status): boolean {
        return this.value === rhs.value;
    }
}
